source/Halibut.Tests/Transport/AuthorizedTcpConnectionsLimiterFixture.cs

```diff
@@ -0,0 +1,228 @@
+
+            //this should throw
+            Action x = () => limiter.ClaimAuthorizedTcpConnection(new Uri("poll://abc"));
+
+            // Assert
+            x.Should().Throw<ActiveTcpConnectionsExceededException>();
+        }
+        
+        [Test]
+        public void CompletedLeasesAreRemovedFromTheCount()
+        {
+            // Arrange
+            const int limit = 3;
+            var subscription = new Uri("poll://abc");
+            var limiter = new ActiveTcpConnectionsLimiter(new HalibutTimeoutsAndLimits
+            {
+                MaximumActiveTcpConnectionsPerPollingSubscription = limit
+            });
+
+            // Act
+            limiter.ClaimAuthorizedTcpConnection(subscription);
+            limiter.ClaimAuthorizedTcpConnection(subscription);
+            
+            //this will decrement the current count in the dispose
+            using ( limiter.ClaimAuthorizedTcpConnection(subscription))
+            {
+            };
+
+            //this should not throw
+            Action x = () => limiter.ClaimAuthorizedTcpConnection(subscription);
+
+            // Assert
+            x.Should().NotThrow<ActiveTcpConnectionsExceededException>();
+        }
+        
+        [Test]
+        public void DoesNotLimitConcurrentConnectionsForDifferentSubscriptions()
+        {
+            // Arrange
+            const int limit = 3;
+            var subscription1 = new Uri("poll://abc");
+            var subscription2 = new Uri("poll://dev");
+            var limiter = new ActiveTcpConnectionsLimiter(new HalibutTimeoutsAndLimits
+            {
+                MaximumActiveTcpConnectionsPerPollingSubscription = limit
+            });
+
+            // Act
+            limiter.ClaimAuthorizedTcpConnection(subscription1);
+            limiter.ClaimAuthorizedTcpConnection(subscription1);
+            limiter.ClaimAuthorizedTcpConnection(subscription1);
+
+            //this should not throw
+            Action x = () => limiter.ClaimAuthorizedTcpConnection(subscription2);
+
+            // Assert
+            x.Should().NotThrow<ActiveTcpConnectionsExceededException>();
+        }
+        
+        [Test]
+        public async Task ShouldHandleMultiThreading()
+        {
+            // Arrange
+            const int limit = 10;
+            var subscription = new Uri("poll://abc");
+            var limiter = new ActiveTcpConnectionsLimiter(new HalibutTimeoutsAndLimits
+            {
+                MaximumActiveTcpConnectionsPerPollingSubscription = limit
+            });
+            
+            // Capture how many claims fail with the exception
+            var failures = 0;
+
+            // Act
+            var tasks = new List<Task>();
+            for (var i = 0; i < 20; i++)
+            {
+                var x = i;
+                tasks.Add(Task.Run(() =>
+                {
+                    try
+                    {
+                        limiter.ClaimAuthorizedTcpConnection(subscription);
```

Comment:
nit: Ideally loop in here as well e.g. for loop 100 times around the try.

Reply:
yeah, good idea!